### PR TITLE
Implement API changes (use traits)

### PR DESCRIPTION
**Pull request by axelchalon**
PR based on https://github.com/tomusdrw/ethabi used by https://github.com/paritytech/sol-rs/ (we can theoretically update the sol-rs dependency to be paritytech/ethabi once this is merged). Part of https://github.com/paritytech/sol-rs/issues/6.

Reference: https://gist.github.com/tomusdrw/e8b6b3fa448a9935b35858e1aa6ceaa9 (PR lets us use the syntax in the gist)

Introduces breaking syntax changes. Some notes:
- Replaces #66 : the closures inside `.call(closure)` and `.transact(closure)` return an IntoFuture. `.call(closure)` and `.transact(closure)` return a Future.
- `contract.constructor.transact(closure)` returns a future of the decoded address of the deployed contract (returned as bytes by the closure). `contract.functions().someFunction(arg).transact(closure)` currently returns a future of (), although I imagine we will want later the closure to return the transaction receipt as bytes and have `.transact(closure)` return the decoded receipt. To do so, we only need the non‑constructor `EthabiFunction`s in the generated code to have the associated type `Output = TransactionReceipt` and the `output` function should be a call to a shared util function to decode transaction receipts. The outputs for non-constant functions, like `transfer` in eip20.abi, are currently still accessible with `outputs().transfer(output_bytes)`
- **help wanted**: cf `let transfer_logs = contract.events().transfer().create_filter(c).call(&caller)?.collect::<Vec<eip20::logs::Transfer>>(); // Read events (same pattern, just passing parse_log as decoder)` in the gist: not sure how to make `.call` work with `TopicFilter`s since I assume the closure would need to take a `TopicFilter` as input (and return a `Result<Vec<RawLog>,_>`); while the Call trait takes `Bytes` as input... we could make another trait TopicFilterCall, wdyt? Or should we transform `Call<Out>` into `Call<In, Out>` ?

Let me know if there's anything that should be improved.

Example gist of generated contract code for Validators.abi: https://gist.github.com/axelchalon/0743285a9b5e2fc8c2414292f0cfb6d8
for eip20.abi: https://gist.github.com/axelchalon/8649a3ed8fe89db5731272b456809e62

@tomusdrw @debris

### CODE REVIEW

File: ethabi/src/lib.rs [hunk 1]
```diff
@@ -63,3 +66,87 @@ pub type Uint = ethereum_types::U256;
...
+	R: futures::IntoFuture<Item=Bytes, Error=String>,
+    F: FnOnce(Bytes) -> R
+{
+    type Result = Box<futures::Future<Item=Out, Error=Error>>;
+
+    fn call<D: 'static>(self, input: Bytes, output_decoder: D) -> Self::Result
+		where D: FnOnce(Bytes) -> Result<Out>
+	{
+		Box::new(
+			(self)(input).into_future().map_err(|e: String| Error::with_chain(Error::from(e), ErrorKind::CallError)).and_then(output_decoder)
```
[the remark this answers is not in the document]
Comment (axelchalon) [reply ethabi/src/lib.rs]: Yes, totally makes sense. Didn't you mean `E: Into<ethabi::Error>`? i.e. anything we can then convert to an ethabi::Error, for example a String

[the remark this answers is not in the document]
Comment (tomusdrw) [reply ethabi/src/lib.rs]: Actually I meant `From<ethabi::Error>` os that user is always getting back his own expected error type (that may encode plenty of conditions that went wrong in `call`), but we require that any `ethabi::Error` that may happen on the way (i.e. cannot encode parameters) is also contained in that user error.

So I want to get rid of `String`s as errors all the way, cause it's super lame to use them.
Example:
`Client::call` in parity returns `CallError` that has stuff like: `OutOfGas`, `InvalidStaticCall`, etc. and when doing `call` from ethabi we actually want to get precisely that error back (+ any potential `ethabi::Error`)

[the remark this answers is not in the document]
Comment (axelchalon) [reply ethabi/src/lib.rs]: Got it! Thanks for the explanation. Just committed

File: ethabi/src/errors.rs [hunk 2]
```diff
@@ -27,5 +27,10 @@ error_chain! {
 			description("Call error"),
 			display("Call error"),
 		}
+
+		TransactError {
```
Comment (tomusdrw) [inline ethabi/src/errors.rs]: Seems unused now.

File: ethabi/src/lib.rs [hunk 3]
```diff
@@ -63,3 +66,88 @@ pub type Uint = ethereum_types::U256;
...
+	R: futures::IntoFuture<Item=Bytes, Error=E>,
+	E: From<Error>
+{
+	type Result = Box<futures::Future<Item=Out, Error=E>>;
+
+	fn call<D: 'static>(self, input: Bytes, output_decoder: D) -> Self::Result
+		where D: FnOnce(Bytes) -> Result<Out>
+	{
+		Box::new(
+			(self)(input).into_future().and_then(|output: Bytes| output_decoder(output).map_err(|e: Error| e.into()))
```
Comment (tomusdrw) [inline ethabi/src/lib.rs]:
```
output_decoder(output).map_err(Into::into)
```
or
```
Ok(output_decoder(output)?)
```
should do

**Comment by snd** [comment]
would be great if this got merged (for sol-rs). let me know if i can help in any way

**Comment by axelchalon** [comment]
I think it's good on my side. I'll just merge master into the branch

**Comment by snd** [comment]
@axelchalon could you bump `ethereum-types` to `0.2.2` here https://github.com/paritytech/ethabi/blob/df87d04d5c76b5fed9194c2ae0a7bba1a3928257/ethabi/Cargo.toml#L18. that would help me in making https://github.com/paritytech/sol-rs/pull/31 compile (i can coerce everything else to use `0.2.2` except for this branch (tried patch, tried update precise, ...) resulting in annoying errors like `the trait std::convert::From<types::U256> is not implemented for ethereum_types::uint::U256` because `types::U256` and `ethereum_types::uint::U256` come from different versions of the same crate).

**Comment by tomusdrw** [comment]
Actually latest codebase is using `ethereum-types=0.3` already.

**Comment by snd** [comment]
@tomusdrw great! as long as parity (evm) and this branch use the same version sol-rs should compile. if they differ we'll get `the trait std::convert::From<types::U256> is not implemented for ethereum_types::uint::U256` etc

**Comment by snd** [comment]
is there anything blocking this?

**Comment by snd** [comment]
would love for this to get merged soon. not just for solaris. the `ParseLog` trait is quite useful for the parity-bridge as well!

**Comment by tomusdrw** [comment]
@debris Ping, can we merge this?

**Comment by tomusdrw** [comment]
@debris Ping, ping

**Comment by snd** [comment]
i don't really understand the need for the `DelegateCall` and `Call` traits.

seems to be about providing access to encoded input and access to a function that decodes output.

that's what `ContractFunction` already provides.

what is gained by the indirection through the `DelegateCall` and `Call` traits?

in other words: instead of implementing `Call` i could just have something take a `T: ContractFunction`. what do i gain by implementing `Call`?

i'm sure a usage example would help me understand!

**Comment by tomusdrw** [comment]
@snd So `ContractFunction` trait is implemented by ethabi, for every function. This is sort of orthogonal to this changes, but allows you to write generic code that works for any ethabi function.

Then we have `DelegateCall` trait that we blanket impl for every `ContractFunction`, I think the reason it was done is to be able to make that feature optional, but perhaps we could merge `DelegateCall` with `ContractFunction` actually.

`Call` is rather implemented outside of the `ethabi` crate, think implemented by `web3`, and that's a struct that given byte input can actually pass it to the node somehow and return output bytes that will later be decoded by `ethabi`.

**Comment by snd** [comment]
@tomusdrw the `ContractFunction` trait is super useful! opens up things like https://github.com/paritytech/ethabi/issues/94#issuecomment-392115836.

> Call is rather implemented outside of the ethabi crate, think implemented by web3, and that's a struct that given byte input can actually pass it to the node somehow and return output bytes that will later be decoded by ethabi.

couldn't `web3` simply take a `ContractFunction` and have access to the encoded input and to the output decode function?

seems to me that `DelegateCall` and `Call` are all about enabling a callback `encoded_input -> raw_output` and having the machinery automatically provide `encoded_input` to it and automatically decode `raw_output`. however by simply taking a `ContractFunction` `f`one can do `f.encoded()`to get `encoded_input` and call `f.output(raw_output)` to get the decoded output.

**taking a `ContractFunction` instead of using the `DelegateCall` and `Call` seems much simpler to me!**

at the moment i can't think of a case where one would need `DelegateCall` and `Call` to accomplish something one can't accomplish by simply taking a `ContractFunction`. an example for its usefulness would be great.

**Comment by tomusdrw** [comment]
@snd Very good point, indeed `DelegateCall/Call` could be as well implemented outside of ethabi library as an extension trait (still with blanket impl for `ContractFunction`).

Obviously the traits are not required but I think that:
```
contract.functions().some_fun(5).call(x)
```
looks better than:
```
call(contract.functions().some_fun(5), x)
```

hence the extension traits.

But if we skip including `DelegateCall` and `Call` in this PR (and for instance move them to web3), we won't have `futures` dependency which is really awesome. @axelchalon do you have some time to remove them from this PR?

**Comment by snd** [comment]
@tomusdrw glad i was able to get the idea across : )

i agree that the chain call syntax looks better.
in my opinion that doesn't justify the complexity added by the `DelegateCall` and `Call` trait.

consider this:
```rust
let web3 = Web3::new(transport);
let contract_connection = web3.contract_at(address);

let future = contract_connection.call(ForeignBridge::default()
    .functions()
    .has_authority_signed_side_to_main(authority_address, message_hash));

let has_signed: bool = try_ready!(future.poll());
```

(inspired by your comment here https://github.com/paritytech/ethabi/issues/94#issuecomment-391498224)

it doesn't require `DelegateCall`/`Call` (which makes it generally easier to understand) and i slightly prefer it over the chain call syntax.

it would be cool to have this merged soon so i can use it in the bridge and sol-rs.
fine with merging it as is and possibly redoing `DelegateCall` / `Call` in a seperate PR.

@axelchalon what are your thoughts on this? happy to help removing the traits.

**Comment by tomusdrw** [comment]
Ok, cool. @snd let's proceed with that - if you have time please update Axel's branch and I'll merge it as soon as it's ready. It's too long overdue already.

**Comment by axelchalon** [comment]
I agree with both of you! :+1:

**Comment by axelchalon** [comment]
I'll remove the traits from this PR

**Comment by axelchalon** [comment]
done.
just for the record, there's a regression because we had introduced `call` on the contract functions in https://github.com/paritytech/ethabi/pull/62, and now it's removed.

**Comment by snd** [comment]
thanks! @axelchalon

**Comment by snd** [comment]
@axelchalon i'm using this branch in the bridge refactoring branch already.

however now i'm getting a ton of:

```
the trait `ethabi::ContractFunction` is not implemented for `contracts::foreign::functions::Signature`
```

`contracts::foreign` is generated through `use_contract!` from `ethabi-derive`.
using this branch for all versions of `ethabi`, `ethabi-derive`, etc.

am i doing something wrong?

**Comment by axelchalon** [comment]
Yes, might be a bit confusing; basically we have, for example for `use_contract!` on the eip20 contract with 'Approve' method:
- functions::Approve, used internally for easy access to the ethabi::Function for Approve
- `Eip20.functions().approve(spender, value)` that returns an instance of `ApproveWithInput`, which implements ContractFunction. (the constructor of `ApproveWithInput` uses functions::Approve under the hood to encode the input)

This being said, I think the user-facing functions will always be the *WithInput ones, returned by `Eip20.functions().function_name(arguments)`. not sure how you came to use the `ethabi::Function`s themselves?

anyway I agree that the name ContractFunction is a bit misleading in this case. maybe the problem is solved if we discourage from using the functions::* directly? or can they still be somehow useful for the user? my mind's a bit fuzzy

**Comment by tomusdrw** [comment]
@debris just add `contract.outputs().<func_name>(...)` maybe without changing the current API?

**Comment by tomusdrw** [comment]
@snd I spoke with @debris a bit about all the use cases and backstories. We have come up with some ideas how to improve the API and make it more consistent. I think his current grumble is about API inconsistency (there are actually at least 3 ways to do the same thing). @debris is going to draft a proposal to change the API soonish.

**Comment by snd** [comment]
great. it could use more consistency. it would be helpful if there were exactly one way to do each thing and if all generated functionality were part of traits. currently there are several things one can't write generic code over since each generated type gets an individual impl. looking forward to the draft

**Pull request merged**